modules/storage/storage-account/main.bicep

```diff
@@ -86,75 +86,54 @@ type cors = {
   ]
 }
 
+@description('The blob service properties for blob soft delete.')
 type deleteRetentionPolicy = {
   allowPermanentDelete: bool
-  days: deleteRetentionPolicyDays?
+  @minValue(1)
+  @maxValue(365)
+  days: int?
   enabled: bool
 }
 
-@minValue(1)
-@maxValue(365)
-type deleteRetentionPolicyDays = int
-
 @description('Toggle to enable or disable versioning for Blob service of the Storage Account. Used only if enableBlobService is set to true.')
-type isBlobVersioningEnabled = bool
+type isVersioningEnabled = bool
 
+@description('The blob service property to configure last access time based tracking policy.')
 type lastAccessTimeTrackingPolicy = {
   blobType: [
     'string'
   ]
   enable: bool
   name: 'AccessTimeTracking'
-  trackingGranularityInDays: trackingGranularityInDays?
+  trackingGranularityInDays: 1?
 }
 
-type trackingGranularityInDays = 1
-
+@description('The blob service property to configure last access time based tracking policy.')
 type restorePolicy = {
-  days: restorePolicyDays?
+  @minValue(1)
+  @maxValue(365)
+  days: int?
   enabled: bool
 }
 
-@minValue(1)
-@maxValue(365)
-type restorePolicyDays = int
-
 type blobServiceContainerProperties = {
-  defaultEncryptionScope: defaultEncryptionScope?
-  denyEncryptionScopeOverride: denyEncryptionScopeOverride?
-  immutableStorageWithVersioning: immutableStorageWithVersioning?
-  publicAccess: blobContainerPublicAccess?
-}
-
-@description('Allowed values are \'Blob\', \'Container\' or \'None\'')
-type blobContainerPublicAccess = string
-
-type defaultEncryptionScope = string
-
-type denyEncryptionScopeOverride = string
-
-type immutableStorageWithVersioning = {
-  enabled: bool
+  defaultEncryptionScope: string?
+  denyEncryptionScopeOverride: string?
+  immutableStorageWithVersioning: {
+    enabled: bool
+  }?
+  @allowed['Blob', 'Container', 'None']
+  publicAccess: string?
 }
 
 type roleAssignmentsArray = {
-  description: roleAssignmentDescription?
-  roleDefinitionIdOrName: roleDefinitionIdOrName?
-  principalIds: principalIds?
-  principalType: principalType?
-  resourceType: resourceType?
+  description: string?
+  roleDefinitionIdOrName: string?
+  principalIds: string[]?
+  principalType: string?
+  resourceType: string?
```

Comment:
I'm not sure if there are any style guidelines for Bicep types yet, but I prefer the approach you take in this PR. All types are purely structural, so nothing would prevent you from assigning a `resourceType` param something expecting a `principalType`. They're both strings, so the assignment wouldn't generate any diagnostics.

Reply:
I also tried to nest it MORE, but did not feel it was better.

```
@description('The properties of a storage account’s Blob service.')
type blobServiceProperties = {
  @description('The blob service properties for change feed events.')
  changeFeed:  {
    enabled: bool
    @minValue(1)
    @maxValue(146000)
    retentionInDays: int?
  }?

  containerDeleteRetentionPolicy: containerDeleteRetentionPolicy?

  @description('Specifies CORS rules for the Blob service. You can include up to five CorsRule elements in the request. If no CorsRule elements are included in the request body, all CORS rules will be deleted, and CORS will be disabled for the Blob service.')
  cors: {
    corsRules: [
      {
        allowedHeaders: [
          'string'
        ]
        allowedMethods: [
          'string'
        ]
        allowedOrigins: [
          'string'
        ]
        exposedHeaders: [
          'string'
        ]
        maxAgeInSeconds: int
      }
    ]
  }?
  deleteRetentionPolicy: deleteRetentionPolicy?
  isVersioningEnabled: isVersioningEnabled?
  lastAccessTimeTrackingPolicy: lastAccessTimeTrackingPolicy?
  restorePolicy: restorePolicy?
}

```